pkg/cache/rediscache/hashcache.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rediscache
+
+import (
+	"time"
+
+	redigo "github.com/gomodule/redigo/redis"
+
+	"github.com/pipe-cd/pipe/pkg/cache"
+	"github.com/pipe-cd/pipe/pkg/redis"
+)
+
+type RedisHashCache struct {
+	redis redis.Redis
+	ttl   uint
+	key   string
+}
+
+func NewHashCache(redis redis.Redis, key string) *RedisHashCache {
+	return &RedisHashCache{
+		redis: redis,
+		key:   key,
+	}
+}
+
+func NewTTLHashCache(redis redis.Redis, ttl time.Duration, key string) *RedisHashCache {
+	return &RedisHashCache{
+		redis: redis,
+		ttl:   uint(ttl.Seconds()),
+		key:   key,
+	}
+}
+
+func (r *RedisHashCache) Get(k interface{}) (interface{}, error) {
+	conn := r.redis.Get()
+	defer conn.Close()
+	reply, err := conn.Do("HGET", r.key, k)
+	if err != nil {
+		if err == redigo.ErrNil {
+			return nil, cache.ErrNotFound
+		}
+		return nil, err
+	}
+	if reply == nil {
+		return nil, cache.ErrNotFound
+	}
+	if err, ok := reply.(redigo.Error); ok {
+		return nil, err
+	}
+	return reply, nil
+}
+
+func (r *RedisHashCache) Put(k interface{}, v interface{}) error {
+	conn := r.redis.Get()
+	defer conn.Close()
+	_, err := conn.Do("HSET", r.key, k, v)
+	if r.ttl != 0 {
+		_, err = conn.Do("EXPIRE", r.key, r.ttl)
+	}
+	return err
+}
+
+func (r *RedisHashCache) Delete(k interface{}) error {
+	conn := r.redis.Get()
+	defer conn.Close()
+	_, err := conn.Do("HDEL", r.key, k)
+	return err
+}
+
+func (r *RedisHashCache) GetAll() ([]interface{}, error) {
+	conn := r.redis.Get()
+	defer conn.Close()
+	reply, err := redigo.StringMap(conn.Do("HGETALL", r.key))
```

Comment:
I think we can not use StringMap in this case.
Because [HGETALL](https://redis.io/commands/hgetall)  is returning an array that containing both field and value.
So we have to use `Array` function to parse the reply. https://pkg.go.dev/github.com/gomodule/redigo/redis#Values

One more point is that instead of returning `[]interface{}` maybe we should return `map[interface{}]interface{}`.

Reply:
If we adopt `map[interface{}]interface{}` as return value, then the current `redigo.StringMap` is good enough since its convert reply to `map[string]string`, how do you think about that 🤔

Reply:
I got it. Agree about that. 👍

Reply:
Adopted returning `map[interface{}]interface{}` idea on https://github.com/pipe-cd/pipe/pull/2190/commits/d7e89e094441c89ffa445e720054a7519205404d 🙆‍♀️ 
It's look much better now, thx 👍